go/tendermint/abci/mux.go

```diff
@@ -363,13 +363,47 @@ func (mux *abciMux) InitChain(req types.RequestInitChain) types.ResponseInitChai
 	return resp
 }
 
+func (s *ApplicationState) InHaltEpoch() bool {
+	epoch, err := s.GetEpoch()
+	if err != nil {
+		s.logger.Error("InHaltEpoch: failed to get current epoch",
+			"err", err,
+		)
+		return false
+	}
+
+	return epoch == epochtime.EpochTime(s.haltEpochHeight)
+}
+
+func (s *ApplicationState) AfterHaltEpoch() bool {
+	epoch, err := s.GetEpoch()
+	if err != nil {
+		s.logger.Error("AfterHaltEpoch: failed to get current epoch",
+			"err", err,
+		)
+		return false
+	}
+	return epoch > epochtime.EpochTime(s.haltEpochHeight)
+}
+
 func (mux *abciMux) BeginBlock(req types.RequestBeginBlock) types.ResponseBeginBlock {
 	blockHeight := mux.state.BlockHeight()
 	mux.logger.Debug("BeginBlock",
 		"req", req,
 		"block_height", blockHeight,
 	)
 
+	if mux.state.InHaltEpoch() {
+		mux.logger.Info("BeginBlock: in halt epoch, emitting empty block.")
+		return types.ResponseBeginBlock{}
+	}
+
+	if mux.state.AfterHaltEpoch() {
+		mux.logger.Info("BeginBlock: halt epoch over, trigginring a halt")
+		// XXX: Ideally there would be a nicer way to do this in tendermint.
```

Comment:
Shouldn't this trigger a graceful shutdown somehow?  We have code that brings the node down cleanly on SIGTERM/SIGINT...